### PR TITLE
Explore Metrics: Update history breadcrumb tooltips

### LINKED ISSUE #90746: Explore Metrics: Add more description to trail breadcrumbs
Around 24% of explore metric users interact with the trail history, however there is limited information in the tooltip to help users navigate through the dots.

This aims to provide more information for users to effectively use the trail breadcrumbs

Image

Figma file: https://www.figma.com/design/eL7u2QrvLykWRLeUELiNNy/%5BGA%5D-Explore-Metrics---Master-File?node-id=718-44076&t=SvV55zI1cpKhJvA9-1

(UX ticket: https://github.com/grafana/grafana/issues/89506)

### PR DESCRIPTION
**Pull request by itsmylife**
**What is this feature?**

Display more information in the history step. 

**Why do we need this feature?**

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/90746

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by itsmylife** [comment]
About filter removal, I just need to know if the filter is removed or added. Right now there is no way to detect that change. So in the next iteration I'll think something about this. 

- `:` I'll add those. 
- That's a bit hard to detect since both step type is `metric`. I'll see what I can do. If that's too hard, can we keep as it is right now?

**Comment by catherineymgui** [comment]
> That's a bit hard to detect since both step type is metric

Okay, let me know what you can do! If it's too hard to change we can do smth like

`Metric:
Selected <name of metric>`

vs 

`Metric:
Metric select page`

?

**Comment by bohandley** [comment]
Code looks great but UI-wise I have similar questions as Catherine.

For label changes, adding a label says "Filter applied: <filter>" and removing a filter says "Filter applied." This seems confusing.

Screenshot 2024-07-23 at 5 02 48 PM
Screenshot 2024-07-23 at 5 03 05 PM


For choosing a metric is says "Metric selected: <metric>" and going back to the original metric select page it says "Metric selected." 

Screenshot 2024-07-23 at 5 00 32 PM
Screenshot 2024-07-23 at 5 01 39 PM

I would try to differentiate those messages!

File: public/app/features/trails/DataTrailsHistory.tsx [hunk 1]
```diff
@@ -1,43 +1,66 @@
...
 
 export type TrailStepType = 'filters' | 'time' | 'metric' | 'start';
+
+const filterSubst = ` $2 `;
+const filterPipeRegex = /(\|)(=|=~|!=|>|<|!~)(\|)/g;
+const stepDescriptionMap: Record<TrailStepType, string> = {
+  start: 'Start of history',
+  metric: 'Metric selected',
+  filters: 'Filter applied',
+  time: 'Time range changed',
```
Comment (bohandley) [inline public/app/features/trails/DataTrailsHistory.tsx:58]: I think we need to add cases for metric removed and filter remved here for those specific message.

**Review by bohandley — COMMENTED** [review]
Code looks good, just concerned about the removal of metric and filters. Those tooltip messages seem confusing with the others.

File: public/app/features/trails/DataTrailsHistory.tsx (hunk 1 above)
Comment (itsmylife) [reply public/app/features/trails/DataTrailsHistory.tsx:58]: I'll do that in the second iteration. Please see this https://github.com/grafana/grafana/pull/90825#issuecomment-2245503932

Also metric selected message should show the selected metric as well. I think there is a little bug that is not showing it.

Comment (bohandley) [reply public/app/features/trails/DataTrailsHistory.tsx:58]: The image I shared was for the workflow "Select a new metric," so not a bug, but applying the code to a new case 😃 

1. Click "Select a new metric"
Screenshot 2024-07-24 at 8 29 55 AM

2. The tooltip shows "Metric selected"
Screenshot 2024-07-24 at 8 30 10 AM

Comment (catherineymgui) [reply public/app/features/trails/DataTrailsHistory.tsx:58]: I suggested this copy (see comment) if we cannot separate the select a new metric from other metric select actions.
https://github.com/grafana/grafana/pull/90825#issuecomment-2245602939

**Comment by yangkb09** [comment]
seconding @catherineymgui & @bohandley's comments, but otherwise LGTM! the added description to the trail breadcrumbs will be helpful

File: public/app/features/trails/DataTrailsHistory.tsx [hunk 2]
```diff
@@ -1,43 +1,66 @@
...
   currentStep: number;
   steps: DataTrailHistoryStep[];
+  filtersApplied: string[];
 }
 
 export function isDataTrailsHistoryState(state: SceneObjectState): state is DataTrailsHistoryState {
   return 'currentStep' in state && 'steps' in state;
 }
 
+export function isDataTrailHistoryFilter(filter?: SceneObjectUrlValue): filter is string[] {
```
Comment (yangkb09) [inline public/app/features/trails/DataTrailsHistory.tsx:36]: the `is` syntax is new to me - cool!

Comment (itsmylife) [reply public/app/features/trails/DataTrailsHistory.tsx:36]: Here is a nice detailed article about all the type guards. https://camchenry.com/blog/typescript-type-guards

File: public/app/features/trails/DataTrailsHistory.tsx [hunk 3]
```diff
@@ -74,6 +75,8 @@ export class DataTrailHistory extends SceneObjectBase<DataTrailsHistoryState> {
       // We always want to ensure in initial 'start' step
       this.addTrailStep(trail, 'start');
 
+      console.log('_onActivate');
```
Comment (bohandley) [inline public/app/features/trails/DataTrailsHistory.tsx:78]: Console! 😄

File: public/app/features/trails/DataTrailsHistory.tsx [hunk 4]
```diff
@@ -47,15 +47,16 @@ export interface DataTrailHistoryStep {
...
 
 const filterSubst = ` $2 `;
 const filterPipeRegex = /(\|)(=|=~|!=|>|<|!~)(\|)/g;
 const stepDescriptionMap: Record<TrailStepType, string> = {
   start: 'Start of history',
-  metric: 'Metric selected',
-  filters: 'Filter applied',
-  time: 'Time range changed',
+  metric: 'Metric selected:',
+  metric_page: 'Metric select page',
```
Comment (bohandley) [inline public/app/features/trails/DataTrailsHistory.tsx:57]: Love it!

**Review by bohandley — APPROVED** [review]
LGTM! 🚀

File: public/app/features/trails/DataTrailsHistory.tsx (hunk 2 above)
Comment (yangkb09) [reply public/app/features/trails/DataTrailsHistory.tsx:36]: thanks for sharing - i bookmarked!

**Pull request merged**